plugins/filter/listify.py

```diff
@@ -0,0 +1,152 @@
+# Copyright: (c) 2017, Ramses Smeyers <rsmeyers@cisco.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {"metadata_version": "1.1", "status": ["preview"], "supported_by": "certified"}
+
+DOCUMENTATION = r"""
+    name: aci_listify
+    short_description: Flattens the nested dictionaries representing the ACI model data.
+    description:
+      - This filter flattens and transforms the input data into a list.
+      - See the Examples section below.
+    options:
+      data:
+        description: This option represents the ACI model data which is a list of dictionaries or a dictionary with any level of nesting data.
+        type: raw
+        required: True
+      keys:
+        description: Comma separated keys of type string denoting the ACI objects.
+        required: True
+"""
+
+EXAMPLES = r"""
+- name: Set vars
+  ansible.builtin.set_fact:
+    data:
+      tenant:
+      - name: ansible_test
+        description: Created using listify
+        app:
+        - name: app_test
+          epg:
+          - name: web
+            bd: web_bd
+          - name: app
+            bd: app_bd
+        bd:
+        - name: bd_test
+          subnet:
+          - name: 10.10.10.1
+            mask: 24
+            scope: private
```

Comment:
How do we deal with this list value, which I think could be [private, shared]?

```
 - name: 10.10.10.1
        mask: 24
        scope: private, shared
```

or would that be something like:

```
 - name: 10.10.10.1
        mask: 24
        scope: 
            - private
            - shared
```

Would the second scenario also be covered in the listify use-case?

Reply:
Hi, we've been using a fork (→ [velotiger/ansible-role-aci-model](https://github.com/velotiger/ansible-role-aci-model)) of the original filter aci_listify and in our inventory we keep such list values as CSV, e.g. `scope: private,shared` (no space after comma). This works very well at least for payloads of module cisco.aci.aci_rest. Therefore I suppose there's no need to support such constructs in the object tree. Besides, a list in a dict is one level down the object tree. Example:
```
        - name: app_test
          epg:
          - name: web
            bd: web_bd
          - name: app
            bd: app_bd
```
IMHO `scope` in your example is almost like `epg` here. It suppose further logic would be required in the filter to handle lists.

Reply:
HI @velotiger, 

Apologies my first example had additional type in space. I meant to highlight a key/value pair, which would be entry into the dictionary. 

Indeed my assumption was that the second example would require additional logic to the code. I don't see `scope` in my example as almost the same as `epg`, since it would handle a different scenario. One would have a list of dictionary (key/value pair) items and the other would be a list of strings.

Reply:
Hi @akinross,

yes, epg is a list of dicts and scope is a list of strings; but:
* the filter does not expect strings in lists (such items will neither be added to the cache nor will they work as keys).
* therefore (as you already said) handling a list (as attribute value) of strings requires additional code.
In the end, cisco.aci.bd-subnet sends CSV to APIC, but a list is required as input:
https://github.com/CiscoDevNet/ansible-aci/blob/ca6110c99741e1af4198ee894b10096f66c3c555/plugins/modules/aci_bd_subnet.py#L411-L415

Workarounds:
* use a filter (e.g. [ansible.builtin.split](https://docs.ansible.com/ansible/devel/collections/ansible/builtin/split_filter.html)) in the taskfile.
* use cisco.aci.aci_rest instead of cisco.aci.bd-subnet

Another (ugly) solution might be to split string values in the module beforehand, e.g.
```python
    if scope is not None:
        if isinstance(scope, str):
            scope = scope.split(',')
        if "private" in scope and "public" in scope:
``` 
But I suppose you do not want to adapt all modules where list attributes are used to the listify filter …

Reply:
Hi @akinross,

I applied a quick fix in my extended version of the listify filter to support list values as attributes: https://github.com/velotiger/ansible-role-aci-model/commit/eed6b9ae722d62943cbf44d5c96898f46df706bc

Reply:
@shrsr can you have a look and see if you think this would be something worth adding?

@velotiger Thank you for the contribution, I see you have added code but not created tests for this. Could you add tests for your changes to the test file: https://github.com/CiscoDevNet/ansible-aci/pull/517/files#diff-94d4006fcc459c51abb36ab9d287d79b89ef0c575fe9e8a8e9c08d3d412704a3

Reply:
Hi @akinross,

you find a test case (subnet 20.20.21.1/24) in the example inventory here: https://github.com/velotiger/ansible-role-aci-model/commit/d07999d4b46d410510def263cd3cd12ba53f091b
I also backported my quick fix to the old filter (slightly optimised before).

Reply:
I made a few changes to the code.
@akinross Can you review?
@velotiger Can you please check if the modified code works for you?

Reply:
Hi @shrsr,

I'm sorry to say the code does not work for me:

* cisco.aci.bd_subnet expects a list as arg for scope, however the code delivers CSV (e.g. `tenant_bd_subnet_scope: public,shared`instead of `tenant_bd_subnet_scope: [public, shared]`).
* Using a generator iterator object instead of a list as output has its drawbacks (e.g. a small CLI tool that I wrote for testing requires a list as input) and I suppose the performance gain is very small.  If performance is an issue (which I doubt) I suggest to get rid of intermediate vars and to replace the last loop by a direct key lookup.

Patch set (without optimisations):
```
--- listify.py	2023-11-14 19:38:21
+++ listify_fix.py	2023-11-14 20:15:14
@@ -225,8 +225,7 @@
 
 
 def listify(d, *keys):
-    for result in listify_worker(d, keys, 0, {}, ""):
-        yield result
+    return [result for result in listify_worker(d, keys, 0, {}, "")]
 
 
 def listify_worker(d, keys, depth, cache, prefix):
@@ -239,7 +238,7 @@
                 for k, v in item.items():
                     if isinstance(v, list) and all(isinstance(x, str) for x in v):
                         cache_key = prefix + k
-                        cache_value = ','.join(v)
+                        cache_value = v
                         cache_work[cache_key] = cache_value
                     elif not isinstance(v, (dict, list)):
                         cache_key = prefix + k
```

Reply:
@velotiger 
I tweaked the function a bit so that it matches the original plugin's output and also uses the generator.

As for your first comment on the scope's values, the previous commit was already taking care of scope: [private, shared] and scope: private,shared.